### PR TITLE
Feature rotation3

### PR DESCRIPTION
**Pull request by DForger**
Implemented Rotation3::Rotate(), Rotateion3::InverseRotate(), and Rotation3::FromRotationMatrix()

### CODE REVIEW

File: lib/kinematics/include/kinematics/matrix_types.h [hunk 1]
```diff
@@ -25,6 +25,8 @@ namespace kinematics {
   typedef Eigen::Matrix<double, 7, 7> Matrix7d;
   typedef Eigen::Matrix<double, 8, 8> Matrix8d;
   typedef Eigen::Matrix<double, 9, 9> Matrix9d;
+
```
Comment (karate362) [inline lib/kinematics/include/kinematics/matrix_types.h:28]: If we only use it in the Rotation3::FromRotationMatrix(), there's no need to add a typedef here.

File: lib/kinematics/src/rotation3.cpp [hunk 2]
```diff
@@ -1,6 +1,6 @@
 // Copyright 2016 Chih-Chung Chou
-#include <iostream>
 #include <math.h>
```
Comment (karate362) [inline lib/kinematics/src/rotation3.cpp:2]: Re-order the include files according to the alphabet and the source.
The order is:
1. C/C++ std
2. 3rd party library
3. Our headers

File: lib/kinematics/src/rotation3.cpp [hunk 3]
```diff
@@ -31,6 +31,17 @@ bool Rotation3::FromQuaternion(const Vector4d& quat) {
   return true;
 }
 
+bool Rotation3::FromRotationMatrix(const Matrix3d &Rm) {
+    Quatd quat(Rm);
```
Comment (karate362) [inline lib/kinematics/src/rotation3.cpp]: Examine if the input is a valid rotation matrix, if not, return false.

File: lib/kinematics/src/rotation3.cpp [hunk 4]
```diff
@@ -31,6 +31,17 @@ bool Rotation3::FromQuaternion(const Vector4d& quat) {
   return true;
 }
 
+bool Rotation3::FromRotationMatrix(const Matrix3d &Rm) {
```
Comment (karate362) [inline lib/kinematics/src/rotation3.cpp:34]: There's no need to use the typedef here.

File: lib/kinematics/src/rotation3.cpp [hunk 5]
```diff
@@ -45,4 +56,16 @@ Matrix3d Rotation3::GetRotationMatrix() const {
   return Rm;
 }
 
+Vector3d Rotation3::Rotate(const Vector3d &px) const {
+    Matrix3d Rm = this->GetRotationMatrix();
```
Comment (karate362) [inline lib/kinematics/src/rotation3.cpp]: When you start a new line, only use two spaces.
  Matrix3d Rm = this->GetRotationMatrix();

Use 4 spaces only when you have to split a long sentence into multiple lines.

File: lib/kinematics/src/rotation3.cpp [hunk 6]
```diff
@@ -45,4 +56,16 @@ Matrix3d Rotation3::GetRotationMatrix() const {
...
+
+    return Rm*px;
```
Comment (karate362) [inline lib/kinematics/src/rotation3.cpp]: 2 spaces.

File: lib/kinematics/src/rotation3.cpp [hunk 7]
```diff
@@ -45,4 +56,16 @@ Matrix3d Rotation3::GetRotationMatrix() const {
...
+}
+
+Vector3d Rotation3::InverseRotate(const Vector3d &px) const {
+    Matrix3d Rm = this->GetRotationMatrix();
```
Comment (karate362) [inline lib/kinematics/src/rotation3.cpp]: 2 spaces

File: lib/kinematics/src/rotation3.cpp [hunk 8]
```diff
@@ -45,4 +56,16 @@ Matrix3d Rotation3::GetRotationMatrix() const {
...
+
+    return Rm.transpose()*px;
```
Comment (karate362) [inline lib/kinematics/src/rotation3.cpp]: 2 spaces.

File: lib/kinematics/test/rotation3_test.cpp [hunk 9]
```diff
@@ -8,26 +8,51 @@
 using kinematics::Matrix3d;
 using kinematics::Rotation3;
 using kinematics::Vector4d;
+using kinematics::Vector3d;
```
Comment (karate362) [inline lib/kinematics/test/rotation3_test.cpp:10]: Reorder the using commands according to alphabets.

File: lib/kinematics/test/rotation3_test.cpp [hunk 10]
```diff
@@ -8,26 +8,51 @@
...
```
Comment (karate362) [inline lib/kinematics/test/rotation3_test.cpp:13]: Use double spaces when starting a new sentence of code, as I mentioned above.

File: lib/kinematics/test/rotation3_test.cpp [hunk 11]
```diff
@@ -8,26 +8,51 @@
...
+    Rx.FromQuaternion(quat);
+    Matrix3d Rm = Rx.GetRotationMatrix();
+    EXPECT_TRUE((Rm - Rt).norm() < 1e-6);
+}
+
+TEST(Rotation3Test, TestRotation) {
+    // A ground truth from Octave (https://www.gnu.org/software/octave/)
+    Vector3d v3dSrc;
+    Vector3d v3dTgt, v3dResult;
+    Vector3d v3dInvTgt, v3dInvResult;
```
Comment (karate362) [inline lib/kinematics/test/rotation3_test.cpp]: Name the variables using "snake" style, like v3d_src, v3d_tgt...
On there other hand, we use "Camel" style to name classes, functions, matrices or global constants like
Rotation3Test, InverseRotate()...

File: lib/kinematics/test/rotation3_test.cpp [hunk 12]
```diff
@@ -8,26 +8,51 @@
...
+
+    Rotation3 Rx;
+    Rx.FromRotationMatrix(Rm);
+
+    v3dResult = Rx.Rotate(v3dSrc);
+    v3dInvResult = Rx.InverseRotate(v3dSrc);
+
+    EXPECT_TRUE((v3dResult - v3dTgt).norm() < 1e-6);
+    EXPECT_TRUE((v3dInvResult - v3dInvTgt).norm() < 1e-6);
 }
```
Comment (karate362) [inline lib/kinematics/test/rotation3_test.cpp:59]: Add a test to see if the Rotate(px) has the same result as GetRotationMatrix() \* px.

File: lib/kinematics/src/rotation3.cpp (hunk 2 above)
Comment (DForger) [inline lib/kinematics/src/rotation3.cpp:2]: noted

File: lib/kinematics/include/kinematics/matrix_types.h (hunk 1 above)
Comment (DForger) [inline lib/kinematics/include/kinematics/matrix_types.h:28]: Noted, I will remove this line

File: lib/kinematics/src/rotation3.cpp (hunk 8 above)
Comment (DForger) [inline lib/kinematics/src/rotation3.cpp]: noted

File: lib/kinematics/test/rotation3_test.cpp (hunk 9 above)
Comment (DForger) [inline lib/kinematics/test/rotation3_test.cpp:10]: noted

File: lib/kinematics/test/rotation3_test.cpp [hunk 13]
```diff
@@ -11,48 +11,48 @@ using kinematics::Vector4d;
...
-    Vector4d quat;
-    Matrix3d Rt;
-    quat << 0.18260418, 0.36510837, -0.54771256, 0.73031674;
-    Rt << -0.666703178340727, -0.666666666666667, 0.333260299984754,
+  // A ground truth from http://www.energid.com/resources/quaternion-calculator/
+  Vector4d quat;
+  Matrix3d Rt;
+  quat << 0.18260418, 0.36510837, -0.54771256, 0.73031674;
+  Rt << -0.666703178340727, -0.666666666666667, 0.333260299984754,
             -0.133231088645414, -0.333333333333334, -0.933347934002774,
```
Comment (karate362) [inline lib/kinematics/test/rotation3_test.cpp]: For initializing matrices, it's fine to use the same indentations:
Rt << -0.666703178340727, -0.666666666666667, 0.333260299984754,
           -0.133231088645414, -0.333333333333334, -0.933347934002774,
           ...

File: lib/kinematics/test/rotation3_test.cpp [hunk 14]
```diff
@@ -23,11 +24,35 @@ TEST(Rotation3Test, TestQuaternion) {
...
 }
+
+TEST(Rotation3Test, TestRotation) {
+  // A ground truth from Octave (https://www.gnu.org/software/octave/)
+  Vector3d v3d_src;
+  Vector3d v3d_tgt, v3d_result;
+  Vector3d v3d_inv_tgt, v3d_inv_result;
+  Matrix3d Rm;
+  v3d_src << 123.531, 332.434, 5654.599;
+  v3d_tgt <<  -3464.041646, 3458.651009, 2852.790003;
```
Comment (karate362) [inline lib/kinematics/test/rotation3_test.cpp]: Remove the extra space after v3d_tgt <<

File: lib/kinematics/test/rotation3_test.cpp (hunk 9 above)
Comment (karate362) [inline lib/kinematics/test/rotation3_test.cpp:10]: Re-order the Vector3d and Vector4d.

File: lib/kinematics/src/rotation3.cpp [hunk 15]
```diff
@@ -45,4 +55,16 @@ Matrix3d Rotation3::GetRotationMatrix() const {
   return Rm;
 }
 
+Vector3d Rotation3::Rotate(const Vector3d &px) const {
+  Matrix3d Rm = this->GetRotationMatrix();
+
```
Comment (karate362) [inline lib/kinematics/src/rotation3.cpp]: Remove the empty line.

File: lib/kinematics/src/rotation3.cpp [hunk 16]
```diff
@@ -45,4 +55,16 @@ Matrix3d Rotation3::GetRotationMatrix() const {
...
+  return Rm*px;
+}
+
+Vector3d Rotation3::InverseRotate(const Vector3d &px) const {
+  Matrix3d Rm = this->GetRotationMatrix();
+
```
Comment (karate362) [inline lib/kinematics/src/rotation3.cpp]: Remove the empty line.

File: lib/kinematics/src/rotation3.cpp [hunk 17]
```diff
@@ -31,6 +31,16 @@ bool Rotation3::FromQuaternion(const Vector4d& quat) {
   return true;
 }
 
+bool Rotation3::FromRotationMatrix(const Matrix3d &Rm) {
+  Eigen::Quaterniond quat(Rm);
+  quat_(0) = quat.w();
+  quat_(1) = quat.x();
+  quat_(2) = quat.y();
+  quat_(3) = quat.z();
+
```
Comment (karate362) [inline lib/kinematics/src/rotation3.cpp]: You can write it as: 
quat_ << quat.w(), quat.x(), quat.y(), quat.z();

Comment (DForger) [inline lib/kinematics/src/rotation3.cpp]: noted.

**Pull request merged**